### ISSUE TITLE
How to use a global favicon.ico?

### ISSUE THREAD

**Issue by firestar115** [perwendel/spark#776]
How do you put a favicon.ico at the base of your website? The favicon is located at the root of the resources. Googling it didn't help.

**Comment by firestar115** [comment]
Using

```java
get("/favicon.ico", (req, resp) -> {
    resp.raw().setContentType(String.valueOf(MediaType.ICO));
    IOUtils.copy(Main.class.getResourceAsStream("/favicon.ico"), resp.raw().getOutputStream());
    return resp.body();
});
```

works for accessing `/favicon.ico` but not any other webpage.

**Comment by tipsy** [comment]
Does putting it in your staticfile-root not work?

`staticFiles.location("/public");`
and 
`public/favicon.ico`
?

**Comment by firestar115** [comment]
Now when going to `[::1]:8080/favicon.ico` it downloads favicon.ico, but it still doesn't show up at `[::1]:8080`.

**Comment by tipsy** [comment]
I can't get my browser to even request the file (without putting a <link> tag). Are you getting a 404 for the favicon, or is your browser simply not loading it?

**Comment by firestar115** [comment]
I'm not getting a 404. After putting my `favicon.ico` in `/public`, I try to access `/` and it keeps the "No Favicon" icon, and going to `/favicon.ico` downloads the file. Maybe it doesn't have the output type specified?

**Comment by tipsy** [comment]
![image](https://cloud.githubusercontent.com/assets/1521451/23107363/4c43cd6c-f6fc-11e6-8f41-ed593f9a7172.png) (test with http://realfavicongenerator.net). Seems the icon is available, but none of my browsers are requesting it.. maybe they have changed the old default-behavior of browsers? 🤔
I always specify in a <link> tag myself, I haven't used global favicon in years. Does this work for you normally?

**Comment by firestar115** [comment]
I know it works on Chrome. In my console, Chrome actually requested it earlier, it was in the logs. I don't know if Chrome stopped requesting it now or not... I recorded it, let me upload it quickly.

**Comment by firestar115** [comment]
[https://youtu.be/H9g1swEAKiE](https://youtu.be/H9g1swEAKiE)

Should I restart Chrome?

**Comment by tipsy** [comment]
You're right about the mime-type. You can do `staticFiles.registerMimeType("ico", "image/x-icon");` to add a mapping (but it doesn't change anything here).

**Comment by firestar115** [comment]
There's not a method called "registerMimeType" for staticFiles?

**Comment by tipsy** [comment]
There is if you bump to a newer version 🤔

**Comment by firestar115** [comment]
Chrome now displays `favicon.ico` as an image (not downloading), but not at the top of tabs.

**Comment by tipsy** [comment]
Shows up in the browser tab now for me. Are you sure your browser is loading it? I couldn't get it to load locally, but it worked after I deployed.

**Comment by firestar115** [comment]
Safari is loading it. I'm going to clear everything and try again with Chrome.

**Comment by firestar115** [comment]
Ah! After force-closing Chrome, the favicon shows up! Mind adding

> # Adding a global favicon
> Place your `favicon.ico` in the `/public` folder, and put this code before your mappings:
> ```java
> staticFiles.location("/public");
> staticFiles.registerMimeType("ico", "image/x-icon");
> ```

to your documentation?

**Comment by tipsy** [comment]
Well, it's not something that comes up a lot (first time!), so I don't think it should go in the docs. I merged 7dfd295 so that you won't have to register the mime-type manually in future versions though.
